arbos/arbosState/arbosstate.go

```diff
@@ -324,6 +324,15 @@ func (state *ArbosState) UpgradeArbosVersion(
 			ensure(params.UpgradeToVersion(2))
 			ensure(params.Save())
 
+		case 32, 33, 34:
+			// these versions are left to Orbit chains for custom upgrades.
+
+		case 35:
```

Comment:
@zacshowa @ImJeremyHe I'm a bit confused now. I feel like when we upgrade to version 35 there is nothing that actually changes in the code. Why does that work? I think there should be some kind of conditional logic that only activates the Espresso sovereign sequencer once the version has been upgraded to 35.

Reply:
So far we haven't been doing this. We've been constructing the config such that the `chainConfig.ArbitrumChainParams.EnableEspresso` config flag is set to true when the new node starts. In theory we could set that in the reference to the chain config that is passed into this function, but it hasn't seemed necessary so far.

Perhaps this allows us to start up the node without worrying what the chain config looks like specifically.

Reply:
IIUC currently this means if one deploys the espresso integration and sets the flags but doesn't upgrade the version (yet) the node will already turn on Espresso mode. I think this is not the correct behaviour. I think the node should turn on the Espresso integration only when the ArbOS version if upgraded because the ArbOS upgrade signifies a change in the STF.

I think it might currently work fine because we test on the simple node. So everything gets upgraded at the same time anyway. So I guess we do have to do the upgrade where we have a least a separate validator node to highlight such troublesome behaviour.

This is where a validator would read the new version, from it's own state

https://github.com/EspressoSystems/nitro-espresso-integration/blob/b2109ff1966778823a84ed268dbd0e36ce8ddf3c/arbos/block_processor.go#L399-L409

But I'm currently not sure how a validator or staker would "receive" the information about when exactly to upgrade. 

@zacshowa @ImJeremyHe @Sneh1999

Reply:
https://github.com/EspressoSystems/nitro-espresso-integration/blob/b56dc56a2409ddcb537a81f4aceba9d3772afd8e/staker/staker.go#L428

Staker checks and updates the module root periodically. And the validation info contains the module root, if the module root doesn't match, then the validation will fail and wait for a certain time until the next check. 
Is this what you wondering?

Reply:
@ImJeremyHe So validators create their own block from the message. The block header contains the ArbOS version. So I think in order to create the correct block the validators need to know the ArbOS version. I don't know if they read this from the internal transaction that does the upgrade or from somewhere else.

Reply:
Or maybe if they upgrade the module root that is sufficient because they now use a replay binary where the version is upgraded? I don't think so. The ArbOS version seems to be stored in the state, not in the replay binary.

Reply:
https://github.com/EspressoSystems/nitro-espresso-integration/pull/216/commits/6110eafe58d100ab4c0d863d2311741d2acc8c0e